aspnetcore/mvc/extensibility/app-parts.md

```diff
@@ -0,0 +1,124 @@
+---
+title: Application Parts | Microsoft Docs
+author: ardalis
+description: 
+keywords: ASP.NET Core,
+ms.author: riande
+manager: wpickett
+ms.date: 1/4/2017
+ms.topic: article
+ms.assetid: b355a48e-a15c-4d58-b69c-899963613a98
+ms.technology: aspnet
+ms.prod: aspnet-core
+uid: mvc/extensibility/app-parts
+---
+# Application Parts
+
+By [Steve Smith](http://ardalis)
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/mvc/extensibility/app-parts/sample)
+
+An *Application Part* is a resource where MVC features may be discovered, such as an assembly. *Feature providers* work with application parts to populate the features of an ASP.NET Core MVC app. 
+The main use case for application parts is to allow you to configure your app to discover (or avoid loading) MVC features from an assembly.
+
+## Introducing Application Parts
+
+MVC apps load their features from [application parts](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.applicationparts.applicationpart). In particular, the [AssemblyPart](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.applicationparts.assemblypart#Microsoft_AspNetCore_Mvc_ApplicationParts_AssemblyPart) class represents an application part that is backed by an assembly. You can use these classes to discover and load MVC features, such as controllers, view components, tag helpers, and razor compilation sources. The [ApplicationPartManager](https://docs.microsoft.com/en-us/aspnet/core/api/microsoft.aspnetcore.mvc.applicationparts.applicationpartmanager) is responsible for tracking the application parts and feature providers available to the MVC app. You can interact with the `ApplicationPartManager` in `Startup` when you configure MVC:
+
+<!-- literal_block {"ids": [], "linenos": true, "xml:space": "preserve", "language": "csharp"} -->
+
+```csharp
+// create an assembly part from a class's assembly
+var assembly = typeof(Startup).GetTypeInfo().Assembly;
+services
+	.AddMvc()
+	.AddApplicationPart(assembly);
+
+// OR
+var assembly = typeof(Startup).GetTypeInfo().Assembly;
+var part = new AssemblyPart(assembly);
+services
+    .AddMvc()
+    .ConfigureApplicationPartManager(apm => p.ApplicationParts.Add(part));
+```
+
+By default MVC will search the dependency tree and find controllers (even in other assemblies). To load an arbitrary assembly (for instance, from a plugin that isn't referenced at compile time), you can use an application part.
+
+You can use application parts to *avoid* looking for controllers in a particular assembly or location. You can control which parts (or assemblies) are available to the app by modifying the `ApplicationParts` collection of the `ApplicationPartManager`. The order of the entries in the `ApplicationParts` collection is not important. It is important to fully configure the `ApplicationPartManager` before using anything that requires services these parts have added.
```

Comment:
It is important to fully configure the `ApplicationPartManager` before using the ApplicationPartManager to configure services in the container. For example, you should fully configure the ApplicationPartManager before invoking AddControllersAsServices(). Failing to do so, will mean that controllers in Application Parts added after that method call will not be affected (will not get registered as services) which might have a negative impact on the correctness of your application.